MAINT, ENH: Move matrix_power to linalg and allow higher dimensions.

mhvk opened the pull request:
The docstring already assumed this was the case, and this ensures linalg becomes completely independent of matrixlib (together with #10975, which moves the tests). Furthermore, since linalg allows stacks of matrices generally, there is no reason not to allow it for matrix_power, so that is made possible.

eric-wieser commented:
Any chance you could split this into two commits - one that just moves the function, and the other that adds the higher-dimensionality support?

eric-wieser commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -925,7 +926,9 @@ def test_large_power_trailing_zero(self):
             mz = matrix_power(M, 0)
-            assert_equal(mz, identity(M.shape[0]))
+            expected = np.empty_like(M)
+            expected[...] = identity(M.shape[0])
```
This could do with a comment explaining why we do it in two steps.

mhvk replied:
Ah, yes, good idea. Even better since that function already was defined in the test file! (as `identity_like_generalized`)

mhvk committed:
ENH: Allow stacks of matrices in matrix_power.

Since linalg allows stacks of matrices generally, there is no
reason not to allow it for matrix_power.

mhvk committed:
MAINT: Move matrix_power to linalg

The docstring already assumed it was in linalg, and this ensures
linalg becomes completely independent of matrixlib.

mhvk commented:
Indeed, reviewing without seeing a good diff is difficult; I now split it into two commits.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -596,47 +585,56 @@ def matrix_power(M, n):
+    a = array(a, subok=True, copy=False, ndmin=1)
+    m = a.shape[-1]
+    if a.ndim < 2 or a.shape[-2] != m:
         raise ValueError("input must be a square array")
```
I think we have some helper functions for this kind of check in this file. `_assert2DSquareness` or similar

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -596,47 +585,56 @@ def matrix_power(M, n):
+        z = matmul(z, z)
+        left, use = divmod(left, 2)
+        if use:
+            result = matmul(result, z)
```
Way better without going through a string!

Could consider also writing:
```python
z = a
while True:
    n, use = divmod(n, 2)
    if use:
        if result is None:
            result = z  # optimization - don't initialize result to the identity
        else:
            result = matmul(result, z)
    if not n:
        break

    z = matmul(z, z)
```

which makes it a bit more obvious that two loops are only needed to avoid a multiplication with the identity

mhvk commented on numpy/linalg/linalg.py:
```diff
@@ -532,6 +532,107 @@ def inv(a):
+        return matmul(matmul(a, a), a)
+
+    # Binary decomposition to reduce the number of Matrix multiplications.
+    # Use None to signal these should be taken as identity matrices below.
```
True. How about
```
    # In the loop, z starts as a, and then gets squared repeatedly,
    # multiplied into the result as needed.
```

mhvk replied:
I like it in a way (the `bits` in particular), but feel the separation ends up making it less clear, so I prefer to just stick with what I have.

eric-wieser replied:
Yeah, I couldn't think of a great way to name the repeated matmul - I don't think `iter_call` was it.

I think using `for use in bits(n):` alone would still be an improvement over the current `while` loop - it makes the intent way clearer

mhvk replied:
I've pushed this up as a separate commit, as I'm a bit torn about whether the separation of code and implementation does not in fact make it less clear.  Have a look what you think.

eric-wieser replied:
You could use a local function, which would reduce the separation a lot. I think it's nice to encapsulate both the loop exit and the iteration in that iterator, rather than mix it with the matrix squaring.

New name looks good to me too.

I'll let you make the final call here - either approach is an improvement over what was here before.

mhvk replied:
OK, in the end I think it is simpler just to keep the loop as I had it, but I changed the comment to (hopefully) be clearer.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -532,7 +532,7 @@ def inv(a):
 
 
-def matrix_power(M, n):
+def matrix_power(a, n):
```
In principle changing the argument name is a compatibility break - perhaps one we can get away with , but maybe worth a remark in the release notes. Maybe mention that it now throws `LinalgError` too?

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -563,26 +562,17 @@ def matrix_power(M, n):
     LinAlgError
         If the matrix is not numerically invertible.
```
Should mention that being non-square does this too.

Also, this comment is sort of false - this only happens if it's non-invertable and `n` is negative.

eric-wieser commented:
Just release notes to go here, I think.

Something for a later PR - supporting an array of powers ought to be fairly easy too.

mhvk commented:
OK, added releasse note and changed the text for the `LinAlgError`

eric-wieser reviewed (approved):
Looks great, thanks!

mhvk commented:
Funny to see this in a matrix-deprecation project! But with this, #10975, and #11010, the main deprecation PR should become quite easy.

charris committed:
Merge branch 'master' into linalg-move-matrix-power

charris commented:
Thanks Marten.
